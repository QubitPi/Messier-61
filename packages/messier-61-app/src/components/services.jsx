/*
 * Copyright 2023 Paion Data. All rights reserved.
 */
import React from "react";
import { useTranslation } from "react-i18next";

export const Services = (props) => {
  const { t } = useTranslation();

  return (
    <div id="services" className="text-center">
      <div className="container">
        <div className="section-title">
          <h2>{t("Highlights")}</h2>
          <p>{t("Highlights Tag Line")}</p>
        </div>
        <div className="row">
          <div key="hightlight-1" className="col-md-4">
            {" "}
            <i className="fa fa-github"></i>
            <div className="hightlight-desc">
              <h3>{t("Hightlights 1 Title")}</h3>
              <p>{t("Hightlights 1 Description")}</p>
            </div>
          </div>

          <div key="hightlight-2" className="col-md-4">
            {" "}
            <i className="fa fa-heart"></i>
            <div className="hightlight-desc">
              <h3>{t("Hightlights 2 Title")}</h3>
              <p>{t("Hightlights 2 Description")}</p>
            </div>
          </div>

          <div key="hightlight-3" className="col-md-4">
            {" "}
            <i className="fa fa-cloud"></i>
            <div className="hightlight-desc">
              <h3>{t("Hightlights 3 Title")}</h3>
              <p>{t("Hightlights 3 Description")}</p>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
