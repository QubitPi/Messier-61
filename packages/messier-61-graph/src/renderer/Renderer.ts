/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BaseType, Selection } from "d3-selection";
import { Visualization } from "../Visualization";
import { NodeCaptionLine, NodeModel } from "../models/Node";
import { BORDER_COLOR, BORDER_WIDTH, COLOR, FONT_SIZE, TEXT_COLOR_INTERNAL } from "../models/GraphStyle";
import { RelationshipModel } from "../models/Relationship";

type RendererEventHandler<Datum> = (
  selection: Selection<SVGGElement, Datum, BaseType, unknown>,
  style: Visualization
) => any;

const NODE_RING_RENDERER_NAME = "nodeRing";
const NODE_OUTLINE_RENDERER_NAME = "nodeOutline";
const NODE_CAPTION_RENDERER_NAME = "nodeCaption";

const ARROW_PATH_RENDERER_NAME = "arrowPath";
const RELATIONSHIP_TYPE_RENDERER_NAME = "relationshipType";

export const NODE_RING_STROKE_WIDTH_IN_PX = 8;
export const NO_OP_RENDERER_HANDLER = () => undefined;

export class Renderer<Datum> {
  public name: string;
  public onTick: RendererEventHandler<Datum>;
  public onGraphChange: RendererEventHandler<Datum>;

  constructor({name, onTick, onGraphChange}: {name: string, onTick: RendererEventHandler<Datum>, onGraphChange: RendererEventHandler<Datum>}) {
    this.name = name;
    this.onTick = onTick;
    this.onGraphChange = onGraphChange;
  }
}

const NODE_OUTLINE_RENDERER = new Renderer<NodeModel>({
  name: NODE_OUTLINE_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("circle.b-outline")
      .data(node => [node])
      .join("circle")
      .classed("b-outline", true)
      .attr("cx", 0)
      .attr("cy", 0)
      .attr("r", (node: NodeModel) => node.radius)
      .attr("fill", (node: NodeModel) => visualizaiton.style.forNode(node).get(COLOR))
      .attr("stroke", (node: NodeModel) => visualizaiton.style.forNode(node).get(BORDER_COLOR))
      .attr("stroke-width", (node: NodeModel) => visualizaiton.style.forNode(node).get(BORDER_WIDTH))
  }
});

const NODE_CAPTION_RENDERER = new Renderer<NodeModel>({
  name: NODE_CAPTION_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("text.caption")
      .data((node: NodeModel) => node.caption)
      .join("text")
      // Classed element ensures duplicate data will be removed befoer adding
      .classed("caption", true)
      .attr("text-anchor", 'middle')
      .attr("pointer-events", 'none')
      .attr("x", 0)
      .attr("y", (nodeCaptionLine: NodeCaptionLine) => nodeCaptionLine.baseline)
      .attr(
        "font-size",
        (nodeCaptionLine: NodeCaptionLine) => visualizaiton.style.forNode(nodeCaptionLine.node).get(FONT_SIZE)
      )
      .attr(
        "fill",
        (nodeCaptionLine: NodeCaptionLine) => visualizaiton.style.forNode(nodeCaptionLine.node).get(TEXT_COLOR_INTERNAL)
      )
      .text((nodeCaptionLine: NodeCaptionLine) => nodeCaptionLine.text)
  }
});

/**
 * The hollow ring that shows up around a node when the mouse mouses over a node
 */
const NODE_RING_RENDERER = new Renderer<NodeModel>({
  name: NODE_RING_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    const circles = selection
      .selectAll("circle.ring")
      .data((node: NodeModel) => [node]);

    circles
      .enter()
      .insert("circle", ".b-outline")
      .classed("ring", true)
      .attr("cx", 0)
      .attr("cy", 0)
      .attr("stroke-width", `${NODE_RING_STROKE_WIDTH_IN_PX}px`)
      .attr("r", (node: NodeModel) => node.radius + 4)

    return circles.exit().remove();
  }
});

const ARROW_PATH_RENDERER = new Renderer<RelationshipModel>({
  name: ARROW_PATH_RENDERER_NAME,
  onTick(selection, visualization) {
    return selection
      .selectAll<BaseType, RelationshipModel>("path")
      .attr("d", d => d.arrow!.outline(d.shortCaptionLength ?? 0))
  },
  onGraphChange(selection, visualization) {
    return selection
      .selectAll("path.b-outline")
      .data(relationship => [relationship])
      .join("path")
      .classed("b-outline", true)
      .attr("fill", (relationship: RelationshipModel) => visualization.style.forRelationship(relationship).get(COLOR))
      .attr("stroke", "none")
  }
});

const RELATIONSHIP_TYPE_RENDERER = new Renderer<RelationshipModel>({
  name: RELATIONSHIP_TYPE_RENDERER_NAME,
  onTick(selection, visualizaiton) {
    return selection
      .selectAll<BaseType, RelationshipModel>("text")
      .attr("x", relationship => relationship?.arrow?.midShaftPoint?.x ?? 0)
      .attr("y", relationship => (relationship?.arrow?.midShaftPoint?.y ?? 0) + parseFloat(visualizaiton.style.forRelationship(relationship).get(FONT_SIZE)) / 2 - 1)
      .attr("transform", relationship => {
        if (relationship.naturalAngle < 90 || relationship.naturalAngle > 270) {
          // Calibrate text orientation. For example,
          // Dragging a node from 1st quadrant to the 2nd quadrant will change the text view to upside-down, making the
          // text hard to read. Rotating the text 180 degress around midshaft point will turn it back to the normal-view
          return `rotate(180 ${relationship?.arrow?.midShaftPoint?.x ?? 0} ${relationship?.arrow?.midShaftPoint?.y ?? 0})`
        } else {
          return null;
        }
      })
      .text(relationship => relationship.shortCaption ?? "");
  },
  onGraphChange(selection, visualizaiton) {
    return selection
      .selectAll("text")
      .data(relationship => [relationship])
      .join("text")
      .attr("text-anchor", "middle")
      .attr("pointer-events", "none")
      .attr("font-size", relationship => visualizaiton.style.forRelationship(relationship).get(FONT_SIZE))
      .attr("fill", relationship => visualizaiton.style.forRelationship(relationship).get(`text-color-${relationship.captionLayout}`));
  }
});

export const NODE_RENDERERS = [NODE_OUTLINE_RENDERER, NODE_CAPTION_RENDERER, NODE_RING_RENDERER];
