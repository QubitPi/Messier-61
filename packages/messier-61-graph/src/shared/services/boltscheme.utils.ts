// Copyright 2023 Paion Data. All rights reserved.
import { INSECURE_SCHEMES, SECURE_SCHEMES } from 'shared/modules/app/appDuck'

const BOLT_DIRECT_SCHEME = 'bolt'
const BOLT_ROUTING_SCHEME = 'neo4j'

export const isNonSupportedRoutingSchemeError = (e: {
  code: string
  message: string
}) =>
  e.code === 'ServiceUnavailable' &&
  (e.message.includes('Could not perform discovery') ||
    e.message.includes('routing'))

export const isNonRoutingScheme = (url = '') =>
  typeof url === 'string' && url.startsWith(`${BOLT_DIRECT_SCHEME}://`)

export const toNonRoutingScheme = (url: string) =>
  typeof url === 'string' &&
  `${BOLT_DIRECT_SCHEME}${getSchemeFlag(url)}://${stripQueryString(
    stripScheme(url)
  )}`

export const getScheme = (url: string) => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = url.split('://')
  return scheme
}

export const stripScheme = (url: string) => {
  const [_scheme, ...rest] = (url || '').split('://')
  if (!rest || !rest.length) {
    return _scheme
  }
  return rest.join('://')
}

export const boltUrlsHaveSameHost = (host1: string, host2: string): boolean =>
  stripScheme(stripQueryString(host1)) === stripScheme(stripQueryString(host2))

export const stripQueryString = (url: string) => {
  const [_host] = (url || '').split('?')
  return _host
}

export const isSecureBoltScheme = (url: string) => {
  if (url && !url.includes('://')) {
    return false
  }
  const [scheme] = (url || '').split('://')
  if (!scheme) {
    return false
  }
  return scheme.endsWith('+s') || scheme.endsWith('+ssc')
}

export const getSchemeFlag = (url = '') => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return ''
  }
  return `+${scheme.split('+').pop()}`
}
const stripSchemeFlag = (url: string) => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return scheme
  }
  return scheme.split('+')[0]
}

const toggleSchemeSecurity = (url: string) => {
  if (url && !url.includes('://')) {
    return url
  }
  if (isSecureBoltScheme(url)) {
    return stripSchemeFlag(url)
  }
  return `${getScheme(url)}+s`
}

export const toggleSchemeRouting = (url = '') => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return url
  }

  if (url.startsWith(BOLT_ROUTING_SCHEME)) {
    return url.replace(BOLT_ROUTING_SCHEME, BOLT_DIRECT_SCHEME)
  }
  if (url.startsWith(BOLT_DIRECT_SCHEME)) {
    return url.replace(BOLT_DIRECT_SCHEME, BOLT_ROUTING_SCHEME)
  }
  return url
}

export const boltToHttp = (boltHost: string): string => {
  const withOutScheme = stripScheme(boltHost)
  if (SECURE_SCHEMES.some(scheme => boltHost.startsWith(scheme))) {
    return `https://${withOutScheme}`
  }
  if (INSECURE_SCHEMES.some(scheme => boltHost.startsWith(scheme))) {
    return `http://${withOutScheme}`
  }
  return boltHost
}

export const generateBoltUrl = (
  allowedSchemes: string[],
  url: string,
  fallbackScheme?: string
) => {
  const rewrites: Record<string, string> = {
    'bolt+routing://': `${BOLT_ROUTING_SCHEME}://`
  }

  if (!url || typeof url !== 'string') {
    if (
      allowedSchemes &&
      fallbackScheme &&
      allowedSchemes.includes(fallbackScheme)
    ) {
      return `${fallbackScheme}://`
    }
    const scheme = allowedSchemes
      ? allowedSchemes[0]
      : fallbackScheme || 'neo4j'
    return `${scheme}://`
  }

  // Rewrite any alias. Break on first hit.
  for (const candidate in rewrites) {
    if (url.startsWith(candidate)) {
      url = `${rewrites[candidate]}${stripScheme(url)}`
      break
    }
  }

  const scheme = getScheme(url)

  // We accept all schemes
  if (!allowedSchemes) {
    if (scheme) {
      return url
    }
    return `${BOLT_ROUTING_SCHEME}://${url}`
  }

  if (!scheme && fallbackScheme && allowedSchemes.includes(fallbackScheme)) {
    return `${fallbackScheme}://${url}`
  }

  // Input scheme allowed
  if (allowedSchemes.includes(scheme)) {
    return url
  }

  // Scheme not allowed, can we toggle encryption to allow it?
  if (scheme) {
    const toggledSecurityScheme = toggleSchemeSecurity(url)
    if (allowedSchemes.includes(toggledSecurityScheme)) {
      return `${toggledSecurityScheme}://${stripScheme(url)}`
    }
  }

  // Either no scheme entered or we can't guess an allowed, let's override it with something default
  const defaultScheme =
    fallbackScheme && allowedSchemes.includes(fallbackScheme)
      ? fallbackScheme
      : allowedSchemes[0]
  return `${defaultScheme}://${stripScheme(url)}`
}
