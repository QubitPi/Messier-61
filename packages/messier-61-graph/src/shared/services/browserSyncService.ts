/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/database'

const appIsInitialized = () => firebase.apps.length && firebase.apps.length > 0

export const authenticate = (dataToken: any, onSignout: any = null) => {
  return firebase
    .auth()
    .signInWithCustomToken(dataToken)
    .then(a => {
      firebase.auth().onAuthStateChanged(user => {
        if (!user || !user.uid) {
          onSignout && onSignout()
        }
      })
      return a
    })
}

export const initialize = (config: any) => {
  if (appIsInitialized()) {
    return
  }

  return firebase.initializeApp(config)
}

export const status = () => {
  return firebase.database().ref('.info/connected')
}

export const getResourceFor = (userId: any) => {
  return firebase.database().ref(`users/${userId}`)
}

export const syncResourceFor = (userId: any, key: any, value: any) => {
  const userRef = firebase.database().ref(`users/${userId}`)
  userRef.child(key).set(value)
}

export const setupUser = (userId: any, initialData: any) => {
  firebase.database().ref(`users/${userId}`).set(initialData)
}

export const signOut = () => {
  if (appIsInitialized()) {
    firebase.auth().signOut()
  }
}
