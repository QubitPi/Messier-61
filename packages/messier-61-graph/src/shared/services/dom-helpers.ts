// Copyright 2023 Paion Data. All rights reserved.

const addClass = (node: any, className: any) => {
  if (!(node instanceof HTMLElement && typeof className === 'string')) {
    return
  }

  // normalize node class name
  const nodeClassName = ` ${node.className} `
  if (nodeClassName.indexOf(` ${className} `) === -1) {
    node.className += (node.className ? ' ' : '') + className
  }
}

const prependIcon = (element: any, classname: string, onClick: () => void) => {
  const icon = document.createElement('i')
  addClass(icon, classname)
  icon.setAttribute('style', 'padding-right:4px')
  element.insertBefore(icon, element.firstChild)

  if (onClick) {
    icon.onclick = e => {
      // prevent populating the editor as well
      e.stopPropagation()
      onClick()
    }
  }
}

export { addClass, prependIcon }
