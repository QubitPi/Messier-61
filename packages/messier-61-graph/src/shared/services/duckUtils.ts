// Copyright 2023 Paion Data. All rights reserved.

export const hydrate = (initialState: any, state: any) => {
  if (!state) return state
  return state.hydrated ? state : { ...initialState, ...state, hydrated: true }
}

export const dehydrate = (state: any) => {
  if (state) {
    delete state.hydrated
  }
  return state
}
