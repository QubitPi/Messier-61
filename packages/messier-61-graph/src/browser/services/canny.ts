/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ReactNode } from 'react'

import { canUseDOM } from 'services/utils'

export const cannyOptions = {
  appID: '601d00fb7a41e3035f75e1e8',
  position: 'right',
  align: 'top'
}
export const CANNY_FEATURE_REQUEST_URL = 'https://feedback.neo4j.com/browser'

export interface CannyOptions {
  appID: string
  position: string
  align: string
}
declare global {
  interface Window {
    Canny?: { (command: string, options?: CannyOptions): void }
    IsCannyLoaded?: boolean
    attachEvent?: typeof window.addEventListener
  }
}

const CannySDK = {
  init: (): Promise<Event> =>
    new Promise(function (resolve, reject) {
      // Code obtained from Canny.io; See: https://developers.canny.io/install

      if (typeof window.Canny === 'function') {
        return
      }

      const canny = function (...args: [string, CannyOptions?]) {
        canny.q.push(args)
      }

      canny.q = [] as [string, CannyOptions?][]
      window.Canny = canny

      function loadCanny() {
        if (document.getElementById('canny-jssdk')) {
          return
        }
        const firstScriptElement = document.getElementsByTagName('script')[0]
        const scriptElement = document.createElement('script')
        scriptElement.type = 'text/javascript'
        scriptElement.async = true
        scriptElement.src = 'https://canny.io/sdk.js'
        scriptElement.onerror = reject
        scriptElement.onload = resolve
        scriptElement.addEventListener('error', reject)
        scriptElement.addEventListener('load', resolve)
        firstScriptElement?.parentNode?.insertBefore(
          scriptElement,
          firstScriptElement
        )
      }

      if (document.readyState === 'complete') {
        loadCanny()
      } else if (window.attachEvent) {
        window.attachEvent('onload', loadCanny)
      } else {
        window.addEventListener('load', loadCanny, false)
      }
    })
}

export class CannyLoader extends Component {
  componentDidMount(): void {
    CannySDK.init()
      .then(() => {
        window.IsCannyLoaded = true
        window.Canny && window.Canny('initChangelog', cannyOptions)
      })
      .catch(() => {
        window.IsCannyLoaded = false
      })
  }

  shouldComponentUpdate(): boolean {
    return false
  }

  componentWillUnmount(): void {
    if (canUseDOM()) {
      window.Canny && window.Canny('closeChangelog')

      delete window.IsCannyLoaded
      delete window.Canny
    }
  }

  render(): ReactNode {
    return null
  }
}
