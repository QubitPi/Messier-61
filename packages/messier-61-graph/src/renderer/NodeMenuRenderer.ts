import { BaseType, Selection } from "d3-selection";
import { Arc, DefaultArcObject, arc as d3Arc } from "d3-shape";
import { NODE_DOUBLE_CLICKED } from "../event-handler/GraphEventHandlerModel";
import { NodeModel } from "../models/Node";
import { NODE_RING_STROKE_WIDTH_IN_PX, NO_OP_RENDERER_HANDLER, Renderer } from "./Renderer";
import { Visualization } from "../Visualization";
import { CONTEXT_MENU_ITEM } from "../VizItem";

/*
 * Copyright Jiaqi Liu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const DONUT_REMOVE_NODE_RENDERER_NAME = "donutRemoveNode";
const DONUT_EXPAND_NODE_RENDERER_NAME = "donutExpandNode";
const DONUT_UNLOCK_NODE_RENDERER_NAME = "donutUnlockNode";

const ITEM_INDEX_MAP = new Map<string, number>([
  [DONUT_REMOVE_NODE_RENDERER_NAME, 1],
  [DONUT_EXPAND_NODE_RENDERER_NAME, 2],
  [DONUT_UNLOCK_NODE_RENDERER_NAME, 3],
]);

const MENU_ANGLE = (2 * Math.PI) / ITEM_INDEX_MAP.size;

const donutExpandNode = new Renderer<NodeModel>({
  name: DONUT_EXPAND_NODE_RENDERER_NAME,
  onTick: NO_OP_RENDERER_HANDLER,
  onGraphChange(selection, visualizaiton) {
    return createMenuItem(selection, visualizaiton, NODE_DOUBLE_CLICKED, 2, NODE_EXPAND_CLASS_NAME, [-8, -10]);
  },
});

export const EXPAND_COLLAPSE_ICON_KEY = "Expand / Collapse";
export const UNLOCK_ICON_KEY = "Unlock";
export const REMOVE_ICON_KEY = "Remove";

export const NODE_EXPAND_CLASS_NAME = "expand-node";

/**
 * Attatch node menues, 
 * @param selection 
 * @param visualizaiton 
 * @param eventType 
 * @param item 
 * @param className 
 * @param position 
 * @param svgIconKey 
 * @param tooltip 
 */
function createMenuItem(
  selection: Selection<SVGGElement, NodeModel, BaseType, unknown>,
  visualizaiton: Visualization,
  eventType: string,
  item:
    | typeof DONUT_REMOVE_NODE_RENDERER_NAME
    | typeof DONUT_EXPAND_NODE_RENDERER_NAME
    | typeof DONUT_UNLOCK_NODE_RENDERER_NAME,
  className: string,
  position: [number, number],
  svgIconKey: typeof EXPAND_COLLAPSE_ICON_KEY | typeof UNLOCK_ICON_KEY | typeof REMOVE_ICON_KEY,
  tooltip: string
) {
  const tab = selection
    .selectAll(`path.${className}`)
    .data(getSelectedNode)
    .join("path")
    .classed(className, true)
    .classed(CONTEXT_MENU_ITEM, true)
    .attr("d", (node) => drawArc(node.radius, item)());

  const rawSvgIcon = 
}

function getSelectedNode(node: NodeModel) {
  return node.selected ? [node] : [];
}

const DEFAULT_WIDTH = 30;

/**
 * Returns a **function** that draws a menu option around a node using [D3 arc](https://github.com/d3/d3-shape#arcs).
 * 
 * When the return value of this function is used as `someFunction` in `.attr("d", someFunction)`, it is very important
 * to remember that the 2nd argument of `.attr()` must either be
 * [constants or functions](https://www.d3indepth.com/selections/#updating-selections-with-functions), therefore we
 * must end the call with an extra parenthesis, i.e.
 * 
 * ```typescript
 * .attr("d", (node) => drawArc(node.radius, item)());
 * ```
 * 
 * Note the last `()` after the `drawArc(node.radius, item)`, We are calling the function returned by this method.
 * i.e. `fn()`, where `fn = drawArc(node.radius, item)`
 * 
 * @param nodeRadius The radius of the enclusing node
 * @param item The menu type/name to draw arc for. Can only be one of the
 * {@link !DONUT_REMOVE_NODE_RENDERER_NAME}/{@link !DONUT_EXPAND_NODE_RENDERER_NAME}/{@link !DONUT_UNLOCK_NODE_RENDERER_NAME}
 * @param arcWidth The difference between outer arc radius and inner ard radius. Defaults to {@link !DEFAULT_WIDTH}.
 *
 * @returns a [D3 arc](https://github.com/d3/d3-shape#arcs) instance
 */
function drawArc(
  nodeRadius: number,
  item:
    | typeof DONUT_REMOVE_NODE_RENDERER_NAME
    | typeof DONUT_EXPAND_NODE_RENDERER_NAME
    | typeof DONUT_UNLOCK_NODE_RENDERER_NAME,
  arcWidth = DEFAULT_WIDTH
): any {
  const startAngle = MENU_ANGLE * ITEM_INDEX_MAP.get(item)!;
  const endAngle = startAngle + MENU_ANGLE;

  const innerRadius = Math.max(nodeRadius + NODE_RING_STROKE_WIDTH_IN_PX, 20);

  return d3Arc()
    .innerRadius(innerRadius)
    .outerRadius(innerRadius + arcWidth)
    .startAngle(startAngle)
    .endAngle(endAngle)
    .padAngle(0.03)!; // https://github.com/d3/d3-shape#arc_padAngle
}
