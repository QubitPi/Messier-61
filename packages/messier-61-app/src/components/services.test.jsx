/*
 * Copyright 2023 Paion Data. All rights reserved.
 */
import renderer from "react-test-renderer";
import { Services } from "./services";

it("services part renders correctly", () => {
  const tree = renderer.create(<Services />).toJSON();
  expect(tree).toMatchSnapshot();
});
