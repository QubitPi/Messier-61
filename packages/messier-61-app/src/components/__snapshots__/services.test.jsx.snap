// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services part renders correctly 1`] = `
<div
  className="text-center"
  id="services"
>
  <div
    className="container"
  >
    <div
      className="section-title"
    >
      <h2>
        Highlights
      </h2>
      <p>
        Paion Data's External Brain is a hightlyl automated Knowledge Management Tool
      </p>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
         
        <i
          className="fa fa-github"
        />
        <div
          className="hightlight-desc"
        >
          <h3>
            Free & Open Source
          </h3>
          <p>
            Most External Brain features are free and will always be free in the future.
          </p>
        </div>
      </div>
      <div
        className="col-md-4"
      >
         
        <i
          className="fa fa-heart"
        />
        <div
          className="hightlight-desc"
        >
          <h3>
            Machine Learning
          </h3>
          <p>
            We revolutionize the personal management by baking Machine Learning as our knowledge management assistant.
          </p>
        </div>
      </div>
      <div
        className="col-md-4"
      >
         
        <i
          className="fa fa-cloud"
        />
        <div
          className="hightlight-desc"
        >
          <h3>
            Cloud Native
          </h3>
          <p>
            You can host your own External Brain instance on your own infrastructure or simply use our service online. Register today to start a new knowledge journey!
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
